db/schema_applier.cc

```diff
@@ -785,8 +785,31 @@ static future<> do_merge_schema(distributed<service::storage_proxy>& proxy, shar
         slogger.debug("Affected tables for keyspace {}: {}", keyspace_name, sel.tables);
     }
 
+    auto merge_keyspace_schemas = [&](auto &&keyspace, auto &&scylla_keyspace) {
+        if (!keyspace.empty() && !scylla_keyspace.empty()) {
+            for (const auto& ks_iter : keyspace) {
+                const auto &ks_name = ks_iter.first;
+                auto ks_rows = keyspace[ks_name]->rows();
+                auto ks_scylla_rows = scylla_keyspace[ks_name]->rows();
+                typeof(ks_rows) out;
+                for (auto &e: ks_rows) {
+                    out.push_back(e);
+                }
+                for (auto &e: ks_scylla_rows) {
+                    out.push_back(e);
+                }
+
+                auto keyspace_merged = make_lw_shared<query::result_set>(keyspace.begin()->second->schema(),
+                                                                         std::move(out));
+                keyspace[ks_name] = keyspace_merged;
+            }
+        }
+    };
```

Comment:
This seems too hacky.

How about we calculate a diff on scylla_keyspaces old vs new, like we do on "keyspaces", and in merge_keyspaces() we augment "altered" with scylla_keyspaces_diff.{entries_only_on_left+entries_only_on_right+entries_differing}?

Reply:
I had it implemented in a similar way already (I'm copy pasting the diff below, couldn't get a link to it in github, even though I overwrote it), but @kbr-scylla had some comments [here](https://github.com/scylladb/scylladb/pull/20852#issuecomment-2379236417) and [here](https://github.com/scylladb/scylladb/pull/20852#issuecomment-2379237640), with which I agree, and I switched to the approach where I'm merging keyspace's metadata. I find the old approach hacky and the new one fitting the design. It was hacky, because we were pushing some extra logic down the call chain and handling corner cases in simple functions with plain logic that just looked at the diff. Current solution does what is should - since keyspace's medatada are placed in 2 separate tables, they should be merged before they can be considered for processing, otherwise we're working on incomplete objects.

```diff
--- a/db/schema_tables.cc
+++ b/db/schema_tables.cc
@@ -1322,6 +1322,7 @@ static future<> do_merge_schema(distributed<service::storage_proxy>& proxy, shar
 
     // current state of the schema
     auto&& old_keyspaces = co_await read_schema_for_keyspaces(proxy, KEYSPACES, keyspaces);
+    auto&& old_scylla_keyspaces = co_await read_schema_for_keyspaces(proxy, SCYLLA_KEYSPACES, keyspaces);
     auto&& old_column_families = co_await read_tables_for_keyspaces(proxy, keyspaces, table_kind::table, affected_tables);
     auto&& old_types = co_await read_schema_for_keyspaces(proxy, TYPES, keyspaces);
     auto&& old_views = co_await read_tables_for_keyspaces(proxy, keyspaces, table_kind::view, affected_tables);
@@ -1333,6 +1334,7 @@ static future<> do_merge_schema(distributed<service::storage_proxy>& proxy, shar
 
     // with new data applied
     auto&& new_keyspaces = co_await read_schema_for_keyspaces(proxy, KEYSPACES, keyspaces);
+    auto&& new_scylla_keyspaces = co_await read_schema_for_keyspaces(proxy, SCYLLA_KEYSPACES, keyspaces);
     auto&& new_column_families = co_await read_tables_for_keyspaces(proxy, keyspaces, table_kind::table, affected_tables);
     auto&& new_types = co_await read_schema_for_keyspaces(proxy, TYPES, keyspaces);
     auto&& new_views = co_await read_tables_for_keyspaces(proxy, keyspaces, table_kind::view, affected_tables);
@@ -1341,6 +1343,8 @@ static future<> do_merge_schema(distributed<service::storage_proxy>& proxy, shar
     auto new_scylla_aggregates = co_await read_schema_for_keyspaces(proxy, SCYLLA_AGGREGATES, keyspaces);
 
     std::set<sstring> keyspaces_to_drop = co_await merge_keyspaces(proxy, std::move(old_keyspaces), std::move(new_keyspaces));
+    bool skip_create = true;
+    std::set<sstring> scylla_keyspaces_to_drop = co_await merge_keyspaces(proxy, std::move(old_scylla_keyspaces), std::move(new_scylla_keyspaces), skip_create);
     auto types_to_drop = co_await merge_types(proxy, std::move(old_types), std::move(new_types));
     co_await merge_tables_and_views(proxy, sys_ks,
         std::move(old_column_families), std::move(new_column_families),
@@ -1371,7 +1375,7 @@ future<lw_shared_ptr<query::result_set>> extract_scylla_specific_keyspace_info(d
     co_return scylla_specific_rs;
 }
 
-future<std::set<sstring>> merge_keyspaces(distributed<service::storage_proxy>& proxy, schema_result&& before, schema_result&& after)
+future<std::set<sstring>> merge_keyspaces(distributed<service::storage_proxy>& proxy, schema_result&& before, schema_result&& after, bool skip_create)
 {
     std::vector<schema_result_value_type> created;
     std::vector<sstring> altered;
@@ -1393,10 +1397,13 @@ future<std::set<sstring>> merge_keyspaces(distributed<service::storage_proxy>& p
         slogger.info("Dropping keyspace {}", key);
         dropped.emplace(key);
     }
-    for (auto&& key : diff.entries_only_on_right) {
-        auto&& value = after[key];
-        slogger.info("Creating keyspace {}", key);
-        created.emplace_back(schema_result_value_type{key, std::move(value)});
+    if (!skip_create) {
+        for (auto &&key: diff.entries_only_on_right) {
+            auto &&value = after[key];
+            slogger.info("Creating keyspace {}", key);
+            slogger.info("Creating keyspace {}", value);
+            created.emplace_back(schema_result_value_type{key, std::move(value)});
+        }
     }
     for (auto&& key : diff.entries_differing) {
         slogger.info("Altering keyspace {}", key);
diff --git a/db/schema_tables.hh b/db/schema_tables.hh
index c64910d3da..09f3c762bf 100644
--- a/db/schema_tables.hh
+++ b/db/schema_tables.hh
@@ -202,7 +202,7 @@ future<> merge_schema(sharded<db::system_keyspace>& sys_ks, distributed<service:
 // of feature_service and schema tables.
 future<> recalculate_schema_version(sharded<db::system_keyspace>& sys_ks, distributed<service::storage_proxy>& proxy, gms::feature_service& feat);
 
-future<std::set<sstring>> merge_keyspaces(distributed<service::storage_proxy>& proxy, schema_result&& before, schema_result&& after);
+future<std::set<sstring>> merge_keyspaces(distributed<service::storage_proxy>& proxy, schema_result&& before, schema_result&& after, bool skip_create = false);
 
 std::vector<mutation> make_create_keyspace_mutations(schema_features features, lw_shared_ptr<keyspace_metadata> keyspace, api::timestamp_type timestamp, bool with_tables_and_types_and_functions = true);
 
```

Reply:
That's not what I'm suggesting. I agree with https://github.com/scylladb/scylladb/pull/20852#issuecomment-2379236417.

I mean this:

```patch
diff --git a/db/schema_applier.cc b/db/schema_applier.cc
index 4f8a397a4e..3d5d6df728 100644
--- a/db/schema_applier.cc
+++ b/db/schema_applier.cc
@@ -246,7 +246,9 @@ static void maybe_delete_schema_version(mutation& m) {
     }
 }
 
-static future<std::set<sstring>> merge_keyspaces(distributed<service::storage_proxy>& proxy, const schema_result& before, const schema_result& after)
+static future<std::set<sstring>> merge_keyspaces(distributed<service::storage_proxy>& proxy,
+                                                 const schema_result& before, const schema_result& after,
+                                                 const schema_result& sk_before, const schema_result& sk_after)
 {
     /*
      * - we don't care about entriesOnlyOnLeft() or entriesInCommon(), because only the changes are of interest to us
@@ -259,22 +261,30 @@ static future<std::set<sstring>> merge_keyspaces(distributed<service::storage_pr
      *   to this node
      */
     auto diff = difference(before, after, indirect_equal_to<lw_shared_ptr<query::result_set>>());
+    auto sk_diff = difference(sk_before, sk_after, indirect_equal_to<lw_shared_ptr<query::result_set>>());
 
     auto& created = diff.entries_only_on_right;
     auto& altered = diff.entries_differing;
     auto& dropped = diff.entries_only_on_left;
 
+    altered.insert(sk_diff.entries_differing.begin(), sk_diff.entries_differing.end());
+    for (auto&& ks : boost::range::join(sk_diff.entries_only_on_right, sk_diff.entries_only_on_left)) {
+        if (!created.contains(ks) && !dropped.contains(ks)) {
+            altered.emplace(ks);
+        }
+    }
+
     auto& sharded_db = proxy.local().get_db();
     for (auto& name : created) {
         slogger.info("Creating keyspace {}", name);
         auto ksm = co_await create_keyspace_from_schema_partition(proxy,
-                schema_result_value_type{name, after.at(name)});
+                schema_result_value_type{name, after.at(name)}, sk_after.at(name));
         co_await replica::database::create_keyspace_on_all_shards(sharded_db, proxy, *ksm);
     }
     for (auto& name : altered) {
         slogger.info("Altering keyspace {}", name);
         auto tmp_ksm = co_await create_keyspace_from_schema_partition(proxy,
-                schema_result_value_type{name, after.at(name)});
+                schema_result_value_type{name, after.at(name)}, sk_after.at(name));
         co_await replica::database::update_keyspace_on_all_shards(sharded_db, *tmp_ksm);
     }
     for (auto& key : dropped) {
@@ -787,6 +797,7 @@ static future<> do_merge_schema(distributed<service::storage_proxy>& proxy, shar
 
     // current state of the schema
     auto&& old_keyspaces = co_await read_schema_for_keyspaces(proxy, KEYSPACES, keyspaces);
+    auto&& old_scylla_keyspaces = co_await read_schema_for_keyspaces(proxy, SCYLLA_KEYSPACES, keyspaces);
     auto&& old_column_families = co_await read_tables_for_keyspaces(proxy, keyspaces, table_kind::table, affected_tables);
     auto&& old_types = co_await read_schema_for_keyspaces(proxy, TYPES, keyspaces);
     auto&& old_views = co_await read_tables_for_keyspaces(proxy, keyspaces, table_kind::view, affected_tables);
@@ -798,6 +809,7 @@ static future<> do_merge_schema(distributed<service::storage_proxy>& proxy, shar
 
     // with new data applied
     auto&& new_keyspaces = co_await read_schema_for_keyspaces(proxy, KEYSPACES, keyspaces);
+    auto&& new_scylla_keyspaces = co_await read_schema_for_keyspaces(proxy, SCYLLA_KEYSPACES, keyspaces);
     auto&& new_column_families = co_await read_tables_for_keyspaces(proxy, keyspaces, table_kind::table, affected_tables);
     auto&& new_types = co_await read_schema_for_keyspaces(proxy, TYPES, keyspaces);
     auto&& new_views = co_await read_tables_for_keyspaces(proxy, keyspaces, table_kind::view, affected_tables);
@@ -805,7 +817,8 @@ static future<> do_merge_schema(distributed<service::storage_proxy>& proxy, shar
     auto new_aggregates = co_await read_schema_for_keyspaces(proxy, AGGREGATES, keyspaces);
     auto new_scylla_aggregates = co_await read_schema_for_keyspaces(proxy, SCYLLA_AGGREGATES, keyspaces);
 
-    std::set<sstring> keyspaces_to_drop = co_await merge_keyspaces(proxy, std::move(old_keyspaces), std::move(new_keyspaces));
+    std::set<sstring> keyspaces_to_drop = co_await merge_keyspaces(proxy, std::move(old_keyspaces), std::move(new_keyspaces),
+                                                                   std::move(old_scylla_keyspaces), std::move(new_scylla_keyspaces));
     auto types_to_drop = co_await merge_types(proxy, std::move(old_types), std::move(new_types));
     co_await merge_tables_and_views(proxy, sys_ks,
         std::move(old_column_families), std::move(new_column_families),
diff --git a/db/schema_tables.cc b/db/schema_tables.cc
index b33943d276..101d754cf6 100644
--- a/db/schema_tables.cc
+++ b/db/schema_tables.cc
@@ -1294,10 +1294,10 @@ std::vector<mutation> make_drop_keyspace_mutations(schema_features features, lw_
  *
  * @param partition Keyspace attributes in serialized form
  */
-future<lw_shared_ptr<keyspace_metadata>> create_keyspace_from_schema_partition(distributed<service::storage_proxy>& proxy, const schema_result_value_type& result)
+future<lw_shared_ptr<keyspace_metadata>> create_keyspace_from_schema_partition(distributed<service::storage_proxy>& proxy,
+                                                                               const schema_result_value_type& result,
+                                                                               lw_shared_ptr<query::result_set> scylla_specific_rs)
 {
-    auto scylla_specific_rs = co_await extract_scylla_specific_keyspace_info(proxy, result);
-
     auto&& rs = result.second;
     if (rs->empty()) {
         throw std::runtime_error("query result has no rows");
diff --git a/db/schema_tables.hh b/db/schema_tables.hh
index adbfe7b099..cf434dca92 100644
--- a/db/schema_tables.hh
+++ b/db/schema_tables.hh
@@ -237,7 +237,7 @@ std::vector<mutation> make_create_keyspace_mutations(schema_features features, l
 
 std::vector<mutation> make_drop_keyspace_mutations(schema_features features, lw_shared_ptr<keyspace_metadata> keyspace, api::timestamp_type timestamp);
 
-future<lw_shared_ptr<keyspace_metadata>> create_keyspace_from_schema_partition(distributed<service::storage_proxy>& proxy, const schema_result_value_type& partition);
+future<lw_shared_ptr<keyspace_metadata>> create_keyspace_from_schema_partition(distributed<service::storage_proxy>& proxy, const schema_result_value_type& partition, lw_shared_ptr<query::result_set> scylla_specific_rs);
 
 std::vector<mutation> make_create_type_mutations(lw_shared_ptr<keyspace_metadata> keyspace, user_type type, api::timestamp_type timestamp);
 
```

As a bonus, we avoid additional query for scylla_keyspaces inside `create_keyspace_from_schema_partition()`.

Your solution is hacky because you create a result_set using schema of "keyspaces" but feed results of a query on "scylla_keyspaces", which has a different schema. This may invoke undefined behavior if result_set is interpreted later.

Reply:
@ptrsmrn Ping ^

Reply:
I adjusted the code a bit and now am looking into CI failures, they seem to be caused by this change.

Reply:
I'm following the suggested code change, hence I'll resolve this comment and continue updating on this issue in the general comments section.